### DISCUSSION TITLE
Difficulties upgrading an old Clang.jl script

### DISCUSSION THREAD

**Discussion by giordano** [JuliaInterop/Clang.jl#414] [Q&A]
I'm trying to upgrade an old Clang.jl script (from v0.13 times) to newer versions of Clang.jl, but I'm facing a few difficulties.  I haven't written this code, so I'm not very familiar with what it's doing.  The code is also non-public, wrapping a proprietary library for proprietary hardware, so not very easy to share.

I'll start from the first one.  The code now looks like this:
```julia
# [...]
clangctx = DefaultContext()
parse_headers!(clangctx, searched_headers, args=[""], includes=includes)
for trans_unit in clangctx.trans_units
    root_cursor = getcursor(trans_unit)
    clang_children = children(root_cursor)
    # do something with `clang_children`....
end
# [...]
```
`getcursor` doesn't seem to be available in newer versions of Clang.jl and I have little clue of what was doing before, so I'm not really sure how I'm supposed to change this at all....  Any suggestions?

**Comment by giordano** [comment]
@Gnimuc bump :slightly_smiling_face:

**Reply by Gnimuc** [reply]
Sorry for the late response. It's getTranslationUnitCursor.

**Reply by Gnimuc** [reply]
@giordano ICYMI

**Reply by giordano** [reply]
Oh, yes, I saw it, but will be able to test it later in the week, I'll get back to you when I try it! :smiley:

**Comment by giordano** [comment]
For the record, in order to upgrade from v0.13 to v0.14 I made the following changes:

1.
    ```julia
    clangctx = DefaultContext()
    parse_headers!(clangctx, ctx.searched_headers, args=[""], includes=includes)
    ```
    to
    ```julia
    clangctx = Generators.create_context(ctx.searched_headers)
    ```
    (not sure what to do with `includes`?)
1.
    ```julia
    getcursor(trans_unit)
    ```
    to
    ```julia
    Clang.getTranslationUnitCursor(trans_unit)
    ```
1.
    ```julia
    filename(child)
    ```
    to
    ```julia
    spelling(child)
    ```
1.
    ```julia
    get_lexical_parent(cursor)
    ```
    to
    ```julia
    Clang.getCursorLexicalParent(cursor)
    ```
1.
    ```julia
    function_args(method)
    ```
    to
    ```julia
    get_function_args(method)
    ```

But now I'm getting an assertion failure `ptr_ref[] != C_NULL` :disappointed: 

As a side note, I understand the need to revise the API from time to time, but detailing the changes needed to upgrade the scripts would be very much valuable for end users :slightly_smiling_face:  This trial-and-error process is not very productive.

**Reply by Gnimuc** [reply]
Sorry for the inconvenience caused by the API changes. 

I almost reworked the package from scratch in v0.14. Back then, the number of users of this package was very low, I found that there were [42 packages](https://github.com/JuliaInterop/Clang.jl/issues/299) on Github that were using Clang.jl's v0.13 API and many of them are already dead. That's why there are no upgrading tips. 

As for the problem, you might find the doc for v0.13 in the following link to be useful:

https://github.com/JuliaInterop/Clang.jl/tree/old-generator#build-a-custom-c-bindings-generator

But the script you are working on is a [custom generator](https://github.com/JuliaInterop/Clang.jl/tree/old-generator#build-a-custom-c-bindings-generator). In most cases, there is no way to upgrade a v0.13 custom generator to v0.14 unless it doesn't use the old `wrap!` function. One example is [libblastrampoline](https://github.com/JuliaLinearAlgebra/libblastrampoline), you could refer to the following PR for how to replace those v0.13 APIs:

https://github.com/JuliaLinearAlgebra/libblastrampoline/pull/41/files#diff-75e6a171860e9b41b32b68d2cfa1a8c56fe01a4e745d3c877ce59703aab57cb3

BTW, Is it possible to directly upgrade the generator script to use the v0.14 API? It should be easier to write a custom generator in v0.14 than in v0.13.
